mavros_extras/src/plugins/wheel_odometry.cpp

```diff
@@ -0,0 +1,525 @@
+		{
+			int iwheel = 0;
+			bool has_param;
+			do {
+				// Build the string in the form "wheelX", where X is the wheel number.
+				// Check if we have "wheelX" parameter.
+				// Indices starts from 0 and should increase without gaps.
+				std::stringstream ss;
+				ss << "wheel" << iwheel++;
+				std::string wheel_name = ss.str();
+				// Check if we have "wheelX" parameter
+				has_param = wo_nh.hasParam(wheel_name);
+
+				if (has_param) {
+					Eigen::Vector2f offset;
+					float radius;
+
+					wo_nh.param(wheel_name+"/x", offset[0], 0.f);
+					wo_nh.param(wheel_name+"/y", offset[1], 0.f);
+					wo_nh.param(wheel_name+"/radius", radius, 0.05f);
+
+					wheel_offset.push_back(offset);
+					wheel_radius.push_back(radius);
+				}
+			}
+			while (has_param);
+
+			// Check for all wheels specified
+			if (wheel_offset.size() < count) {
+				odom_mode = OM::NONE;
+				ROS_WARN_NAMED("wo", "WO: Not all wheels have parameters specified (%lu/%i).", wheel_offset.size(), count);
+			}
+			// Duplicate 1st wheel if only one is available.
+			// This generalizes odometry computations for 1- and 2-wheels configurations.
+			else if (wheel_radius.size() == 1) {
+				wheel_offset.resize(2);
+				wheel_radius.resize(2);
+				wheel_offset[1].x() = wheel_offset[0].x();
+				wheel_offset[1].y() = wheel_offset[0].y() + 1.f; // make separation non-zero to avoid div-by-zero
+				wheel_radius[1] = wheel_radius[0];
+			}
+
+			// Check for non-zero wheel separation (first two wheels)
+			float separation = std::abs(wheel_offset[1].y() - wheel_offset[0].y());
+			if (separation < 1.e-5f) {
+				odom_mode = OM::NONE;
+				ROS_WARN_NAMED("wo", "WO: Separation between the first two wheels is too small (%f).", separation);
+			}
+
+			// Check for reasonable radiuses
+			for (int i = 0; i < wheel_radius.size(); i++) {
+				if (wheel_radius[i] <= 1.e-5f) {
+					odom_mode = OM::NONE;
+					ROS_WARN_NAMED("wo", "WO: Wheel #%i has incorrect radius (%f).", i, wheel_radius[i]);
+				}
+			}
+		}
+
+		// Advertise RPM-s and distance-s
+		if (diagnostics) {
+			rpm_pub = wo_nh.advertise<mavros_msgs::Float32ArrayStamped>("rpm", 10);
+			dist_pub = wo_nh.advertise<mavros_msgs::Float32ArrayStamped>("distance", 10);
+		}
+
+		// Advertize topics
+		if (odom_mode != OM::NONE) {
+			if (twist_send)
+				twist_pub = wo_nh.advertise<geometry_msgs::TwistWithCovarianceStamped>("velocity", 10);
+			else
+				odom_pub = wo_nh.advertise<nav_msgs::Odometry>("odom", 10);
+		}
+		// No-odometry warning
+		else
+			ROS_WARN_NAMED("wo", "WO: No odometry computations will be performed.");
+
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return {
+			make_handler(&WheelOdometryPlugin::handle_rpm),
+			make_handler(&WheelOdometryPlugin::handle_wheel_distance)
+		};
+	}
+
+private:
+	ros::NodeHandle wo_nh;
+
+	ros::Publisher rpm_pub;
+	ros::Publisher dist_pub;
+	ros::Publisher odom_pub;
+	ros::Publisher twist_pub;
+	tf2_ros::TransformBroadcaster tfBroadcaster;
```

Comment:
Why, UAS already have one.

Reply:
https://github.com/mavlink/mavros/blob/master/mavros/include/mavros/mavros_uas.h#L271

Reply:
Yes, forgot to delete.